File: wind-composer-web/src/music/intelligentComposer.ts
import type { CompositionPlan, MelodyNoteDto, RhythmEventDto, WeatherSnapshot } from "../types";
import { clamp } from "../utils";
import { ArrangementEngine } from "./arrangementEngine";
import { BassEngine } from "./bassEngine";
import { FillEngine } from "./fillEngine";
import { LeadEngine } from "./leadEngine";
import { MusicMemory } from "./musicMemory";
import {
  beatMicroDecision,
  gustActionWeights,
  pickGustAction,
} from "./probabilityEngine";
import type { ScaleEngine } from "./scaleEngine";
import { getStyle } from "./styleEngine";
import { TempoEngine, windToTargetBpm } from "./tempoEngine";
import { TransitionEngine } from "./transitionEngine";
import type { WeatherChangeSummary, WeatherMemory } from "./weatherMemory";

export interface EnhanceContext {
  gust: boolean;
  samplePosition: number;
  sampleRate: number;
  windKmh: number;
  personalityHope: number;
}

export class IntelligentComposer {
  private styleName = "Ambient";
  private memory = new MusicMemory();
  private arrangement = new ArrangementEngine();
  private bass: BassEngine;
  private lead: LeadEngine;
  private fills: FillEngine;
  private transitions: TransitionEngine;
  private tempo = new TempoEngine();
  private localTimeStr = "";
  private lastNotice = "";
  private lastChangeSummary: WeatherChangeSummary | null = null;
  private fillProbabilityBoost = 0;
  private phraseNumber = 0;
  private lastBeat = -1;
  private lastMeasure = -1;
  private lastBassPattern: number[] = [];

  constructor(
    private scale: ScaleEngine,
    private weatherMemory: WeatherMemory,
  ) {
    this.bass = new BassEngine(this.memory);
    this.lead = new LeadEngine(this.scale, this.memory);
    this.fills = new FillEngine(this.memory);
    this.transitions = new TransitionEngine(this.memory);
  }

  setStyle(name: string): void {
    this.styleName = name;
    const style = getStyle(name);
    this.tempo.reset((style.bpmMin + style.bpmMax) / 2);
    this.lastBeat = -1;
    this.lastMeasure = -1;
    this.lastBassPattern = [];
  }

  getLocalTime(): string {
    return this.localTimeStr;
  }

  getLastNotice(): string {
    return this.lastNotice;
  }

  getChangeSummary(): WeatherChangeSummary | null {
    return this.lastChangeSummary;
  }

  getFillProbabilityBoost(): number {
    return this.fillProbabilityBoost;
  }

  getBassPattern(): number[] {
    return this.lastBassPattern;
  }

  resetForPlayback(): void {
    this.lastBeat = -1;
    this.lastMeasure = -1;
    this.lastBassPattern = [];
    this.fillProbabilityBoost = 0;
  }

  onWeatherUpdated(snap: WeatherSnapshot, localTimeStr: string): void {
    this.localTimeStr = localTimeStr;
    const prev = this.weatherMemory.push(snap);
    if (!prev) return;
    const summary = this.weatherMemory.changeSummary(prev, snap);
    if (!summary.lines.length) return;
    this.lastChangeSummary = summary;
    this.lastNotice = [...summary.lines, ...summary.musical_hints].join("\n");
    this.fillProbabilityBoost = clamp(this.fillProbabilityBoost + summary.fill_prob_delta, 0, 0.25);
  }

  enhance(plan: CompositionPlan, weather: WeatherSnapshot | null, ctx: EnhanceContext): CompositionPlan {
    const style = getStyle(this.styleName);
    const trend = this.weatherMemory.trend();
    const w = weather;

    const windKmh = w ? trend.avg_wind_kmh + trend.wind_delta * 0.2 : ctx.windKmh;
    const targetBpm = windToTargetBpm(
      windKmh,
      style.bpmMin,
      style.bpmMax,
      trend.wind_delta,
      trend.storm_likelihood,
    );

    const estimateBpm = this.tempo.getBpm();
    const spbEst = (60 / Math.max(estimateBpm, 40)) * ctx.sampleRate;
    const measureEst = Math.floor(Math.floor(ctx.samplePosition / spbEst) / 4);
    plan.tempo_bpm = this.tempo.update(targetBpm, style.bpmMin, style.bpmMax, measureEst, ctx.gust);

    const spb = (60 / Math.max(plan.tempo_bpm, 40)) * ctx.sampleRate;
    const beat = Math.floor(ctx.samplePosition / spb);
    const measure = Math.floor(beat / 4);
    const prevBeat = this.lastBeat;
    const prevMeasure = this.lastMeasure;

    plan.musical_style = this.styleName;
    plan.local_time_str = this.localTimeStr;

    if (measure > prevMeasure) {
      const measureStart = Math.max(1, prevMeasure + 1);
      const measureEnd = Math.min(measure, measureStart + 7);
      for (let m = measureStart; m <= measureEnd; m++) {
        plan.song_section = this.arrangement.onBar(m, plan.energy_curve, trend.storm_likelihood > 0.5);
        if (m > 0 && m % 32 === 0) this.phraseNumber += 1;
      }
    } else {
      plan.song_section = this.arrangement.getState().section;
    }
    plan.phrase_number = this.phraseNumber;

    const extraRhythm: RhythmEventDto[] = [];
    const extraMelody: MelodyNoteDto[] = [];
    const arrState = this.arrangement.getState();

    if (beat > prevBeat) {
      const beatStart = Math.max(0, prevBeat + 1);
      const beatEnd = Math.min(beat, beatStart + 31);
      for (let b = beatStart; b <= beatEnd; b++) {
        const micro = beatMicroDecision(plan.energy_curve);
        if (micro === "hat_ghost") {
          extraRhythm.push({ layer: "hat_ghost", strength: 0.14, is_pulse: false });
        }
        if (ctx.gust && w && b === beat) {
          const gustDelta = w.wind_gust_kmh - w.wind_speed_kmh;
          const weights = gustActionWeights(w.wind_speed_kmh, gustDelta, plan.energy_curve);
          const action = pickGustAction(weights);
          if (action === "fill") extraRhythm.push({ layer: "fill", strength: 0.72, is_pulse: true });
          else if (action === "lead_flourish" && plan.chord) {
            extraMelody.push({
              midi: plan.chord.tones[plan.chord.tones.length - 1],
              velocity: 0.82,
              duration_sec: 0.4,
            });
          } else if (action === "crash") extraRhythm.push({ layer: "crash", strength: 0.88, is_pulse: true });
          else if (action === "reverse_fx") plan.transition_fx = "reverse_crash";
          else if (action === "riser") plan.transition_fx = "noise_riser";
          else if (action === "bass_variation" && plan.chord) {
            extraMelody.push({
              midi: plan.chord.tones[0] - 7,
              velocity: 0.55,
              duration_sec: 0.25,
            });
          }
          plan.weather_hints = [...(plan.weather_hints ?? []), `Gust: ${action}`];
        }
      }
    }

    if (measure > prevMeasure) {
      const measureStart = Math.max(1, prevMeasure + 1);
      const measureEnd = Math.min(measure, measureStart + 7);
      for (let m = measureStart; m <= measureEnd; m++) {
        const tones = plan.chord?.tones ?? [];
        const bassNotes = this.bass.onBar({
          chordTones: tones,
          energy: plan.energy_curve,
          windDirection: w?.wind_direction_deg ?? 0,
          pressureTrend: trend.pressure_delta,
          barInPhrase: m,
          style,
        });
        extraMelody.push(...bassNotes);
        if (bassNotes.length) {
          this.lastBassPattern = bassNotes.map((n) => n.midi);
        }

        const fillProb = style.fillProbability + this.fillProbabilityBoost;
        extraRhythm.push(
          ...this.fills.maybeFill({
            bar: m,
            phraseLength: plan.phrase_length_bars,
            energy: plan.energy_curve,
            stormLikelihood: trend.storm_likelihood,
            fillProbability: fillProb,
          }),
        );
        this.fillProbabilityBoost *= 0.92;

        if (plan.chord?.tones?.length) {
          extraMelody.push(
            ...this.lead.maybeNotes({
              chordTones: plan.chord.tones,
              energy: plan.energy_curve,
              hope: ctx.personalityHope,
              bar: m,
              gust: ctx.gust,
              style,
            }),
          );
        }

        const fx = this.transitions.maybeTransition({
          energy: plan.energy_curve,
          stormLikelihood: trend.storm_likelihood,
          sectionChange: arrState.barsInSection === 0,
          transitionProbability: style.transitionProbability,
          gust: ctx.gust,
        });
        if (fx) {
          plan.transition_fx = fx;
          if (["crash", "impact", "reverse_crash"].includes(fx)) {
            extraRhythm.push({ layer: "crash", strength: 0.78, is_pulse: true });
          }
        }
      }
    }

    this.lastBeat = beat;
    this.lastMeasure = measure;

    if (w) {
      if (w.snowfall_mm > 0.1) plan.brightness = clamp(plan.brightness * 0.9);
      if (w.precipitation_mm > 0.5) plan.percussion = Math.max(plan.percussion, w.precipitation_mm / 8);
      if (w.humidity_pct > 70) plan.reverb_amount = clamp(plan.reverb_amount + 0.1);
      if (w.cloud_cover_pct > 80) plan.reverb_amount = clamp(plan.reverb_amount + 0.06);
      if (w.temperature_c < 5) plan.brightness = clamp(plan.brightness * 0.88);
      if (w.temperature_c > 25) plan.brightness = clamp(plan.brightness + 0.08);
    }

    if (trend.storm_likelihood > 0.4) plan.weather_hints = [...(plan.weather_hints ?? []), "Storm tension"];
    if (trend.calm_trend) plan.weather_hints = [...(plan.weather_hints ?? []), "Calming trend"];
    if (trend.accelerating_wind) plan.weather_hints = [...(plan.weather_hints ?? []), "Wind accelerating"];

    plan.rhythm_events = extraRhythm;
    plan.melody_notes = extraMelody.slice(0, 8);
    plan.bass_notes = extraMelody.filter((n) => n.midi < 52);
    plan.drum_events = extraRhythm;

    return plan;
  }

  getArrangementLayerGains(energy: number): Record<string, number> {
    return this.arrangement.layerGains(getStyle(this.styleName), energy);
  }
}

File: wind-composer/intelligent_composer.py
"""Phase 5 intelligent electronic composer — orchestrates style, weather, engines."""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from datetime import datetime
from typing import List, Optional

from arrangement_engine import ArrangementEngine
from bass_engine import BassContext, BassEngine
from composition_engine import CompositionContext, CompositionPlan
from drum_engine import DrumContext, DrumEngine
from fill_engine import FillContext, FillEngine
from lead_engine import LeadContext, LeadEngine
from melody_engine import MelodyNote
from music_memory import MusicMemory
from probability_engine import beat_micro_decision, gust_action_weights, pick_gust_action
from rhythm_engine import RhythmEvent
from scale_engine import ScaleEngine
from style_engine import MusicalStyle, SongSection, get_style, style_bpm_range
from tempo_engine import TempoEngine, wind_to_target_bpm
from transition_engine import TransitionContext, TransitionEngine
from weather_memory import WeatherMemory, WeatherTrend
from weather.models import WeatherSnapshot
from utils import clamp


@dataclass
class WeatherChangeNotice:
    lines: List[str] = field(default_factory=list)
    musical_hints: List[str] = field(default_factory=list)
    timestamp: float = 0.0


class IntelligentComposer:
  """Extends composition plans with style-aware electronic arrangement."""

  def __init__(self, scale_engine: ScaleEngine) -> None:
    self._scale = scale_engine
    self._memory = MusicMemory()
    self._weather_memory = WeatherMemory(20)
    self._arrangement = ArrangementEngine()
    self._bass = BassEngine(self._memory)
    self._drums = DrumEngine()
    self._lead = LeadEngine(scale_engine, self._memory)
    self._fills = FillEngine(self._memory)
    self._transitions = TransitionEngine(self._memory)
    self._style_name = MusicalStyle.AMBIENT.value
    self._tempo = TempoEngine()
    self._last_notice: Optional[WeatherChangeNotice] = None
    self._local_time_str = ""
    self._rng = random.Random()

  @property
  def style_name(self) -> str:
    return self._style_name

  def set_style(self, name: str) -> None:
    self._style_name = name
    style = get_style(name)
    self._tempo.reset((style.bpm_min + style.bpm_max) / 2.0)

  def get_last_weather_notice(self) -> Optional[WeatherChangeNotice]:
    return self._last_notice

  def on_weather_updated(
    self,
    snap: WeatherSnapshot,
    local_time_str: str,
  ) -> Optional[WeatherChangeNotice]:
    prev = self._weather_memory.push(snap)
    self._local_time_str = local_time_str
    if prev is None:
      return None
    lines = self._weather_memory.change_summary(prev, snap)
    if not lines:
      return None
    trend = self._weather_memory.trend()
    hints: List[str] = ["New Phrase Generated"]
    if trend.storm_likelihood > 0.5:
      hints.append("Fill Probability Increased")
    if trend.accelerating_wind:
      hints.append("Bass Variation Added")
    if trend.pressure_delta < -2:
      hints.append("Tension Building")
    if abs(snap.wind_speed_kmh - prev.wind_speed_kmh) > 5:
      e_delta = (snap.wind_speed_kmh - prev.wind_speed_kmh) / 40.0
      hints.append(f"Energy {e_delta:+.0%}")
    notice = WeatherChangeNotice(lines=lines, musical_hints=hints, timestamp=datetime.now().timestamp())
    self._last_notice = notice
    return notice

  def enhance(
    self,
    ctx: CompositionContext,
    plan: CompositionPlan,
    beat: int,
    measure: int,
    personality_hope: float,
  ) -> CompositionPlan:
    style = get_style(self._style_name)
    bpm_min, bpm_max = style.bpm_min, style.bpm_max
    plan.musical_style = self._style_name

    trend = self._weather_memory.trend()
    w = ctx.weather

    wind_kmh = w.wind_speed_kmh if w else 0.0
    if w:
      wind_kmh = trend.avg_wind_kmh if hasattr(trend, "avg_wind_kmh") else w.wind_speed_kmh
      if hasattr(trend, "wind_delta"):
        wind_kmh += trend.wind_delta * 0.2

    target_bpm = wind_to_target_bpm(
      wind_kmh,
      bpm_min,
      bpm_max,
      trend.wind_delta,
      trend.storm_likelihood,
    )
    plan.tempo_bpm = self._tempo.update(
      target_bpm,
      bpm_min,
      bpm_max,
      measure,
      ctx.gust,
    )

    section = self._arrangement.on_phrase(plan.energy_curve, trend.storm_likelihood > 0.5)
    plan.song_section = section.value
    plan.local_time_str = self._local_time_str

    layer_gains = self._arrangement.layer_gains(style, plan.energy_curve)
    plan.active_instruments = self._arrangement.state.active_layers

    extra_rhythm: List[RhythmEvent] = list(plan.rhythm_events)
    extra_melody: List[MelodyNote] = list(plan.melody_notes)

    # Beat-level drums and micro evolution
    if beat >= 0:
      drum_ctx = DrumContext(
        beat=beat,
        bar=measure,
        energy=plan.energy_curve,
        precipitation=w.precipitation_mm if w else 0,
        snowfall=w.snowfall_mm if w else 0,
        style=style,
        section_energy=self._arrangement.state.section_energy,
      )
      extra_rhythm.extend(self._drums.on_beat(drum_ctx))
      micro = beat_micro_decision(plan.energy_curve, self._rng)
      if micro == "hat_ghost":
        extra_rhythm.append(RhythmEvent("hat", 0.15, False))

    # Bar-level bass and fills
    if measure > 0 and beat % 4 == 0:
      tones = plan.chord.tones if plan.chord else []
      bass_ctx = BassContext(
        chord_tones=tones,
        energy=plan.energy_curve,
        wind_direction=w.wind_direction_deg if w else 0,
        pressure_trend=trend.pressure_delta,
        bar_in_phrase=measure,
        style=style,
      )
      extra_melody.extend(self._bass.on_bar(bass_ctx))

      fill_ctx = FillContext(
        bar=measure,
        phrase_length=plan.phrase_length_bars,
        energy=plan.energy_curve,
        storm_likelihood=trend.storm_likelihood,
        fill_probability=style.fill_probability,
      )
      extra_rhythm.extend(self._fills.maybe_fill(fill_ctx))

    # Lead flourishes
    if plan.chord and plan.chord.tones:
      lead_ctx = LeadContext(
        chord_tones=plan.chord.tones,
        energy=plan.energy_curve,
        hope=personality_hope,
        bar=measure,
        gust=ctx.gust,
        style=style,
      )
      extra_melody.extend(self._lead.maybe_notes(lead_ctx))

    # Gust-weighted actions
    if ctx.gust and w:
      gust_delta = w.wind_gust_kmh - w.wind_speed_kmh
      weights = gust_action_weights(w.wind_speed_kmh, gust_delta, plan.energy_curve)
      action = pick_gust_action(weights, self._rng)
      if action == "fill":
        extra_rhythm.append(RhythmEvent("percussion", 0.7, True))
      elif action == "lead_flourish" and plan.chord:
        extra_melody.append(MelodyNote(midi=plan.chord.tones[-1], velocity=0.8, duration_sec=0.4))
      elif action == "crash":
        extra_rhythm.append(RhythmEvent("percussion", 0.85, True))
      plan.weather_hints.append(f"Gust: {action}")

    # Transitions on section boundaries
    trans_ctx = TransitionContext(
      energy=plan.energy_curve,
      storm_likelihood=trend.storm_likelihood,
      section_change=self._arrangement.state.bars_in_section == 0,
      transition_probability=style.transition_probability,
      gust=ctx.gust,
    )
    fx = self._transitions.maybe_transition(trans_ctx)
    if fx:
      plan.transition_fx = fx
      if fx in ("crash", "impact", "reverse_crash"):
        extra_rhythm.append(RhythmEvent("percussion", 0.75, True))

    # Weather modifiers on harmony density
    if w:
      if w.snowfall_mm > 0.1:
        plan.atmosphere_gain = clamp(plan.atmosphere_gain + 0.15)
        plan.lead_gain *= 0.6
      if w.precipitation_mm > 0.5:
        plan.percussion = max(plan.percussion, w.precipitation_mm / 8)
      if w.humidity_pct > 70:
        plan.reverb_amount = clamp(plan.reverb_amount + 0.12)
      if w.cloud_cover_pct > 80:
        plan.pad_gain = clamp(plan.pad_gain + 0.1)
      if w.temperature_c < 5:
        plan.brightness = clamp(plan.brightness * 0.85)
      if w.temperature_c > 25:
        plan.brightness = clamp(plan.brightness + 0.1)

    plan.rhythm_events = extra_rhythm
    plan.melody_notes = extra_melody[:6]
    plan.drum_events = [e for e in extra_rhythm if e.layer in ("kick", "snare", "hat", "percussion")]
    plan.bass_notes = [n for n in extra_melody if n.midi < 52]

    # Store hints for UI
    if trend.storm_likelihood > 0.4:
      plan.weather_hints.append("Storm tension")
    if trend.calm_trend:
      plan.weather_hints.append("Calming trend")

    return plan

  def reset(self) -> None:
    self._memory.reset()
    self._arrangement = ArrangementEngine()
